test/concurrent.cc

```diff
@@ -0,0 +1,717 @@
+                                });
+                              } else {
+                                callbacks.emplace_back([]() {});
+                              }
+                            })
+                            .interrupt([](auto& context, auto& k) {
+                              if (context == 2) {
+                                k.Stop();
+                              }
+                            }));
+           })
+        | Reduce(
+               std::string(),
+               [](auto& result) {
+                 return Then([&](auto&& value) {
+                   result = value;
+                   return false; // Only take the first element!
+                 });
+               });
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  ASSERT_EQ(2, callbacks.size());
+
+  EXPECT_EQ(
+      std::future_status::timeout,
+      future.wait_for(std::chrono::seconds(0)));
+
+  for (auto& callback : callbacks) {
+    callback();
+  }
+
+  EXPECT_EQ("1", future.get());
+}
+
+TEST(ConcurrentTest, DownstreamDoneOneEventualFail) {
+  std::deque<Callback<>> callbacks;
+
+  auto e = [&]() {
+    return Iterate({1, 2})
+        | Concurrent([&]() {
+             return Map(Eventual<std::string>()
+                            .context(0)
+                            .start([&](auto& context, auto& k, int i) mutable {
+                              context = i;
+                              if (i == 1) {
+                                callbacks.emplace_back([&k]() {
+                                  k.Start("1");
+                                });
+                              } else {
+                                callbacks.emplace_back([]() {});
+                              }
+                            })
+                            .interrupt([](auto& context, auto& k) {
+                              if (context == 2) {
+                                k.Fail("error");
+                              }
+                            }));
+           })
+        | Reduce(
+               std::string(),
+               [](auto& result) {
+                 return Then([&](auto&& value) {
+                   result = value;
+                   return false; // Only take the first element!
+                 });
+               });
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  ASSERT_EQ(2, callbacks.size());
+
+  EXPECT_EQ(
+      std::future_status::timeout,
+      future.wait_for(std::chrono::seconds(0)));
+
+  for (auto& callback : callbacks) {
+    callback();
+  }
+
+  EXPECT_EQ("1", future.get());
+}
+
+TEST(ConcurrentTest, StreamForEach) {
+  auto e = []() {
+    return Iterate({1, 2})
+        | Concurrent([]() {
+             return StreamForEach([](int i) {
+               return Range(i);
+             });
+           })
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*e(), UnorderedElementsAre(0, 0, 1));
+}
```

Comment:
There is not a single comment in here, which for my untrained eye makes most of these tests rather hard to read or reason about. You have a ton of nice comments in the headers. Could I ask you to add a line or two per test?